File: 152-maximum-product-subarray/maximum-product-subarray.c
int maxProduct(int* nums, int numsSize) {
    if (numsSize == 0) return 0;
    int max = nums[0];
    int min = nums[0];
    int res = max;
    for (int i = 1; i < numsSize; i++) {
        int pos = max;
        max = fmax(nums[i], fmax(max * nums[i], min * nums[i]));
        min = fmin(nums[i], fmin(pos * nums[i], min * nums[i]));
        res = fmax(res, max);
    }
    return res;
}

File: 152-maximum-product-subarray/maximum-product-subarray.rs
impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        if nums.is_empty() { return 0; }
        if nums.len() == 1 { return nums[0]; }
        let mut max = nums[0];
        let mut min = nums[0];
        let mut res = max;
        let mut pos;
        for &num in nums.iter().skip(1) {
            pos = max;
            max = num.max(max * num).max(min * num);
            min = num.min(pos * num).min(min * num);
            res = res.max(max);
        }
        res
    }
}